### PR TITLE
Pass configs as paths instead of objects 

### PR DESCRIPTION
**Pull request by padmaia**
# ↪️ Pull Request
Also adds a cli option to specify which config to use

## ✔️ PR Todo

- [ ] Add some tests
- [x] Invalidate correctly when default config is specified as a package name

### CODE REVIEW

File: packages/core/core/src/requests/ParcelConfigRequest.js [hunk 1]
```diff
@@ -73,13 +71,14 @@ export default function createParcelConfigRequest(): ParcelConfigRequest {
         api.invalidateOnFileDelete(filePath);
       }
 
-      if (config.filePath === options.defaultConfig?.filePath) {
+      // TODO: this wouldn't work if default config is a package name
```
Comment (padmaia) [inline packages/core/core/src/requests/ParcelConfigRequest.js:74]: Called this out as a TODO in the description

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 6.40s | -228.00ms |
| Cached | 2.27s | -15.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 277.00ms | -35.00ms 🚀 |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 276.00ms | -37.00ms 🚀 |
| dist/legacy/index.42516b4a.js | 1.10kb | +0.00b | 1.55s | -176.00ms 🚀 |
| dist/modern/index.8a1ced31.js | 1.10kb | +0.00b | 1.54s | -177.00ms 🚀 |
| dist/modern/index.e480862d.css | 77.00b | +0.00b | 1.44s | -137.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 202.00ms | -75.00ms 🚀 |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 201.00ms | -80.00ms 🚀 |
| dist/legacy/index.42516b4a.js | 1.10kb | +0.00b | 400.00ms | -136.00ms 🚀 |
| dist/modern/index.8a1ced31.js | 1.10kb | +0.00b | 399.00ms | -136.00ms 🚀 |
| dist/legacy/index.html | 701.00b | +0.00b | 401.00ms | -136.00ms 🚀 |
| dist/modern/index.html | 701.00b | +0.00b | 399.00ms | -136.00ms 🚀 |
| dist/legacy/index.0e158b6b.css | 77.00b | +0.00b | 400.00ms | -136.00ms 🚀 |
| dist/modern/index.e480862d.css | 77.00b | +0.00b | 399.00ms | -136.00ms 🚀 |

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 27.91s | +11.00ms |
| Cached | 2.37s | -78.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 478.42kb | +0.00b | 188.00ms | -119.00ms 🚀 |
| dist/PermalinkedComment.c85d137d.js | 4.22kb | +0.00b | 179.00ms | -126.00ms 🚀 |
| dist/UserProfile.30e15baf.js | 1.62kb | +0.00b | 179.00ms | -126.00ms 🚀 |
| dist/NotFound.5aa6b041.js | 387.00b | +0.00b | 180.00ms | -125.00ms 🚀 |
| dist/logo.24c8bf9e.png | 274.00b | +0.00b | 136.00ms | -52.00ms 🚀 |

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 2.90m | +2.29s |
| Cached | 5.01s | -268.00ms 🚀 |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/popup.4a66e15e.js | 170.62kb | +0.00b | 34.39s | -19.57s 🚀 |
| dist/media-viewer.1d9322e7.js | 69.09kb | +0.00b | 16.82s | +906.00ms ⚠️ |
| dist/Modal.e3bd77d0.js | 41.40kb | +0.00b | 6.17s | +478.00ms ⚠️ |
| dist/component.59edff1f.js | 30.78kb | +0.00b | 3.11s | +597.00ms ⚠️ |
| dist/component.37867f50.js | 22.51kb | +0.00b | 6.12s | +426.00ms ⚠️ |
| dist/js.48be70a4.js | 16.40kb | +0.00b | 6.12s | +426.00ms ⚠️ |
| dist/dropzone.c8c03b6a.js | 14.36kb | +0.00b | 58.31s | +3.91s ⚠️ |
| dist/component.11476a92.js | 6.05kb | +0.00b | 6.12s | +426.00ms ⚠️ |
| dist/media-viewer.bc661bce.js | 3.92kb | +0.00b | 16.79s | +875.00ms ⚠️ |
| dist/index.393afe59.css | 3.46kb | +0.00b | 1.03m | +3.39s ⚠️ |
| dist/png-chunks-extract.8d15a914.js | 3.41kb | +0.00b | 6.12s | +426.00ms ⚠️ |
| dist/dropzone.2f1a1e17.js | 3.23kb | +0.00b | 31.38s | -19.70s 🚀 |
| dist/Modal.2835744c.js | 3.08kb | +0.00b | 3.11s | -1.44s 🚀 |
| dist/16.d1b8e6f7.js | 2.38kb | +0.00b | 3.11s | +596.00ms ⚠️ |
| dist/card.cadccd96.js | 2.03kb | +0.00b | 17.00s | +947.00ms ⚠️ |
| dist/16.86a8c7d4.js | 1.66kb | +0.00b | 8.18s | +595.00ms ⚠️ |
| dist/status.cccf04be.js | 1.46kb | +0.00b | 7.65s | +397.00ms ⚠️ |
| dist/16.cff410ff.js | 1.38kb | +0.00b | 6.25s | +346.00ms ⚠️ |
| dist/16.c1ed07f8.js | 1.33kb | +0.00b | 8.18s | +595.00ms ⚠️ |
| dist/16.6603f2f3.js | 1.33kb | +0.00b | 6.24s | +344.00ms ⚠️ |
| dist/16.a4b392c2.js | 1.32kb | +0.00b | 3.11s | +613.00ms ⚠️ |
| dist/heading3.93606b9c.js | 1.29kb | +0.00b | 7.66s | +404.00ms ⚠️ |
| dist/16.e60b4d77.js | 1.21kb | +0.00b | 4.65s | +2.15s ⚠️ |
| dist/expand.6889aead.js | 1.16kb | +0.00b | 7.80s | +406.00ms ⚠️ |
| dist/media-card-analytics-error-boundary.6a25b970.js | 1.15kb | +0.00b | 31.36s | -19.72s 🚀 |
| dist/16.2096b9f3.js | 1.14kb | +0.00b | 3.11s | +614.00ms ⚠️ |
| dist/16.f040d62d.js | 1.14kb | +0.00b | 6.26s | +357.00ms ⚠️ |
| dist/16.ecf8a702.js | 1.13kb | +0.00b | 8.18s | +492.00ms ⚠️ |
| dist/16.1b342686.js | 1.13kb | +0.00b | 6.25s | +346.00ms ⚠️ |
| dist/16.2e7f52c6.js | 1.13kb | +0.00b | 6.25s | +346.00ms ⚠️ |
| dist/16.da43aa07.js | 1.13kb | +0.00b | 8.18s | +594.00ms ⚠️ |
| dist/heading2.e219309a.js | 1.11kb | +0.00b | 7.65s | +397.00ms ⚠️ |
| dist/heading1.fa66b375.js | 976.00b | +0.00b | 7.65s | +397.00ms ⚠️ |
| dist/table.b35545be.js | 889.00b | +0.00b | 7.65s | +397.00ms ⚠️ |
| dist/media-picker-analytics-error-boundary.553055f6.js | 837.00b | +0.00b | 34.22s | -16.96s 🚀 |
| dist/media-viewer-analytics-error-boundary.bbe40a02.js | 757.00b | +0.00b | 16.79s | +875.00ms ⚠️ |
| dist/EmojiPickerComponent.a968dafe.js | 535.00b | +0.00b | 12.39s | +1.79s ⚠️ |
| dist/index.html | 119.00b | +0.00b | 1.00m | +59.46s ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.4ca65e68.js | 2.73mb | -2.00b 🚀 | 530.00ms | +79.00ms ⚠️ |
| dist/pdfRenderer.cd742f5e.js | 1.11mb | +0.00b | 528.00ms | +88.00ms ⚠️ |
| dist/editorView.ae59b49f.js | 585.98kb | +0.00b | 527.00ms | +87.00ms ⚠️ |
| dist/popup.4a66e15e.js | 170.62kb | +0.00b | 528.00ms | +88.00ms ⚠️ |
| dist/EmojiPickerComponent.bab709db.js | 97.89kb | +0.00b | 529.00ms | +88.00ms ⚠️ |
| dist/Toolbar.33a83ae3.js | 92.34kb | +0.00b | 527.00ms | +87.00ms ⚠️ |
| dist/media-viewer.1d9322e7.js | 69.09kb | +0.00b | 528.00ms | +88.00ms ⚠️ |
| dist/card.ce67dc35.js | 55.09kb | +0.00b | 528.00ms | +88.00ms ⚠️ |
| dist/card.1007a95f.js | 52.86kb | +0.00b | 527.00ms | +87.00ms ⚠️ |
| dist/Modal.e3bd77d0.js | 41.40kb | +0.00b | 531.00ms | +89.00ms ⚠️ |
| dist/component.59edff1f.js | 30.78kb | +0.00b | 530.00ms | +87.00ms ⚠️ |
| dist/component.37867f50.js | 22.51kb | +0.00b | 531.00ms | +89.00ms ⚠️ |
| dist/DatePicker.cbe44a9b.js | 21.17kb | +0.00b | 529.00ms | +88.00ms ⚠️ |
| dist/EmojiPickerComponent.898d8be6.js | 18.23kb | +0.00b | 529.00ms | +88.00ms ⚠️ |
| dist/js.48be70a4.js | 16.40kb | +0.00b | 531.00ms | +89.00ms ⚠️ |
| dist/dropzone.c8c03b6a.js | 14.36kb | +0.00b | 527.00ms | +87.00ms ⚠️ |
| dist/ui.cee430a3.js | 13.97kb | +0.00b | 528.00ms | +87.00ms ⚠️ |
| dist/smartMediaEditor.7ee0079c.js | 13.95kb | +0.00b | 527.00ms | +87.00ms ⚠️ |
| dist/workerHasher.65b8a1b7.js | 11.78kb | +0.00b | 529.00ms | +88.00ms ⚠️ |
| dist/EmojiTypeAheadComponent.9af6e5fc.js | 8.67kb | +0.00b | 529.00ms | +88.00ms ⚠️ |
| dist/component.11476a92.js | 6.05kb | +0.00b | 531.00ms | +89.00ms ⚠️ |
| dist/card.4ac5c1b2.js | 5.61kb | +0.00b | 528.00ms | +88.00ms ⚠️ |
| dist/media-viewer.bc661bce.js | 3.92kb | +0.00b | 528.00ms | +88.00ms ⚠️ |
| dist/EmojiPickerComponent.624bf3f4.js | 3.58kb | +0.00b | 529.00ms | +88.00ms ⚠️ |
| dist/index.393afe59.css | 3.46kb | +0.00b | 527.00ms | +88.00ms ⚠️ |
| dist/png-chunks-extract.8d15a914.js | 3.41kb | +0.00b | 531.00ms | +89.00ms ⚠️ |
| dist/dropzone.2f1a1e17.js | 3.23kb | +0.00b | 528.00ms | +88.00ms ⚠️ |
| dist/Modal.2835744c.js | 3.08kb | +0.00b | 530.00ms | +87.00ms ⚠️ |
| dist/clipboard.0244488f.js | 2.75kb | +0.00b | 527.00ms | +87.00ms ⚠️ |
| dist/16.d1b8e6f7.js | 2.38kb | +0.00b | 530.00ms | +87.00ms ⚠️ |
| dist/EmojiUploadComponent.88a444df.js | 2.18kb | +0.00b | 529.00ms | +88.00ms ⚠️ |
| dist/card.cadccd96.js | 2.03kb | +0.00b | 528.00ms | +88.00ms ⚠️ |
| dist/ResourcedEmojiComponent.d9a49606.js | 2.00kb | +0.00b | 530.00ms | +79.00ms ⚠️ |
| dist/date.d49e64f8.js | 1.73kb | +0.00b | 530.00ms | +88.00ms ⚠️ |
| dist/16.f40c1329.js | 1.73kb | +0.00b | 530.00ms | +87.00ms ⚠️ |
| dist/images.80270b1f.js | 1.67kb | +0.00b | 530.00ms | +88.00ms ⚠️ |
| dist/16.86a8c7d4.js | 1.66kb | +0.00b | 530.00ms | +87.00ms ⚠️ |
| dist/feedback.8b7c7928.js | 1.64kb | +0.00b | 529.00ms | +88.00ms ⚠️ |
| dist/browser.16c438d4.js | 1.61kb | +0.00b | 527.00ms | +87.00ms ⚠️ |
| dist/workerHasher.9bb425b1.js | 1.54kb | +0.00b | 529.00ms | +88.00ms ⚠️ |
| dist/workerHasher.f66f00c7.js | 1.54kb | +0.00b | 527.00ms | +87.00ms ⚠️ |
| dist/list-number.4632b459.js | 1.46kb | +0.00b | 530.00ms | +88.00ms ⚠️ |
| dist/status.cccf04be.js | 1.46kb | +0.00b | 529.00ms | +88.00ms ⚠️ |
| dist/16.cff410ff.js | 1.38kb | +0.00b | 530.00ms | +87.00ms ⚠️ |
| dist/code.af9ca2b3.js | 1.38kb | +0.00b | 530.00ms | +88.00ms ⚠️ |
| dist/16.c1ed07f8.js | 1.33kb | +0.00b | 531.00ms | +88.00ms ⚠️ |
| dist/16.6603f2f3.js | 1.33kb | +0.00b | 530.00ms | +87.00ms ⚠️ |
| dist/16.a4b392c2.js | 1.32kb | +0.00b | 530.00ms | +87.00ms ⚠️ |
| dist/16.249c644a.js | 1.31kb | +0.00b | 530.00ms | +88.00ms ⚠️ |
| dist/link.d12cc130.js | 1.30kb | +0.00b | 530.00ms | +88.00ms ⚠️ |
| dist/heading6.4c4017db.js | 1.30kb | +0.00b | 529.00ms | +88.00ms ⚠️ |
| dist/heading3.93606b9c.js | 1.29kb | +0.00b | 529.00ms | +88.00ms ⚠️ |
| dist/16.ab5c0088.js | 1.27kb | +0.00b | 530.00ms | +88.00ms ⚠️ |
| dist/emoji.e603e70e.js | 1.23kb | +0.00b | 530.00ms | +88.00ms ⚠️ |
| dist/16.afee1494.js | 1.23kb | +0.00b | 530.00ms | +87.00ms ⚠️ |
| dist/16.e60b4d77.js | 1.21kb | +0.00b | 530.00ms | +87.00ms ⚠️ |
| dist/16.7c7a40c9.js | 1.19kb | +0.00b | 530.00ms | +87.00ms ⚠️ |
| dist/16.06d5a67b.js | 1.18kb | +0.00b | 530.00ms | +88.00ms ⚠️ |
| dist/heading5.dde15cec.js | 1.17kb | +0.00b | 529.00ms | +88.00ms ⚠️ |
| dist/expand.6889aead.js | 1.16kb | +0.00b | 529.00ms | +88.00ms ⚠️ |
| dist/media-card-analytics-error-boundary.6a25b970.js | 1.15kb | +0.00b | 528.00ms | +88.00ms ⚠️ |
| dist/16.2096b9f3.js | 1.14kb | +0.00b | 530.00ms | +87.00ms ⚠️ |
| dist/16.f040d62d.js | 1.14kb | +0.00b | 530.00ms | +87.00ms ⚠️ |
| dist/16.ecf8a702.js | 1.13kb | +0.00b | 530.00ms | +87.00ms ⚠️ |
| dist/16.1b342686.js | 1.13kb | +0.00b | 530.00ms | +87.00ms ⚠️ |
| dist/16.2e7f52c6.js | 1.13kb | +0.00b | 530.00ms | +87.00ms ⚠️ |
| dist/16.da43aa07.js | 1.13kb | +0.00b | 531.00ms | +88.00ms ⚠️ |
| dist/heading2.e219309a.js | 1.11kb | +0.00b | 529.00ms | +88.00ms ⚠️ |
| dist/16.dd07505e.js | 1.10kb | +0.00b | 530.00ms | +87.00ms ⚠️ |
| dist/component.40194190.js | 1.07kb | +0.00b | 530.00ms | +79.00ms ⚠️ |
| dist/mention.b1828447.js | 1.07kb | +0.00b | 529.00ms | +87.00ms ⚠️ |
| dist/heading4.5316f6f2.js | 1.06kb | +0.00b | 529.00ms | +88.00ms ⚠️ |
| dist/16.317b9164.js | 1.05kb | +0.00b | 530.00ms | +88.00ms ⚠️ |
| dist/layout.ac6734f9.js | 1.05kb | +0.00b | 530.00ms | +88.00ms ⚠️ |
| dist/divider.d2be55a1.js | 1.03kb | +0.00b | 530.00ms | +88.00ms ⚠️ |
| dist/quote.f4e09143.js | 1.02kb | +0.00b | 529.00ms | +88.00ms ⚠️ |
| dist/action.cb287eb0.js | 1.00kb | +0.00b | 530.00ms | +88.00ms ⚠️ |
| dist/decision.91f3af14.js | 1010.00b | +0.00b | 530.00ms | +88.00ms ⚠️ |
| dist/panel-warning.68b04096.js | 1008.00b | +0.00b | 529.00ms | +88.00ms ⚠️ |
| dist/list.057cb664.js | 978.00b | +0.00b | 530.00ms | +88.00ms ⚠️ |
| dist/heading1.fa66b375.js | 976.00b | +0.00b | 529.00ms | +88.00ms ⚠️ |
| dist/panel-error.c4b05f6f.js | 906.00b | +0.00b | 529.00ms | +87.00ms ⚠️ |
| dist/panel.9dbc048e.js | 897.00b | +0.00b | 529.00ms | +88.00ms ⚠️ |
| dist/table.b35545be.js | 889.00b | +0.00b | 529.00ms | +88.00ms ⚠️ |
| dist/panel-success.7eb0809a.js | 845.00b | +0.00b | 529.00ms | +88.00ms ⚠️ |
| dist/panel-note.ec8386b7.js | 841.00b | +0.00b | 529.00ms | +88.00ms ⚠️ |
| dist/media-picker-analytics-error-boundary.553055f6.js | 837.00b | +0.00b | 528.00ms | +88.00ms ⚠️ |
| dist/media-card-analytics-error-boundary.964206cc.js | 833.00b | +0.00b | 528.00ms | +88.00ms ⚠️ |
| dist/media-card-analytics-error-boundary.57503222.js | 833.00b | +0.00b | 527.00ms | +87.00ms ⚠️ |
| dist/media-viewer-analytics-error-boundary.bbe40a02.js | 757.00b | +0.00b | 528.00ms | +88.00ms ⚠️ |
| dist/EmojiPickerComponent.a968dafe.js | 535.00b | +0.00b | 529.00ms | +88.00ms ⚠️ |
| dist/simpleHasher.f8e69464.js | 528.00b | +0.00b | 528.00ms | +87.00ms ⚠️ |
| dist/simpleHasher.32d9352e.js | 528.00b | +0.00b | 527.00ms | +87.00ms ⚠️ |
| dist/index.html | 119.00b | +0.00b | 532.00ms | +79.00ms ⚠️ |

</p></details>

<details><summary>Three.js x4 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/271804539956363788)**

**Review by wbinnssmith — APPROVED** [review]

**Comment by EmielM** [comment]
@padmaia Hmm, is dropping support for programmatically passing config objects intentional? I'm giving Parcel v2 a spin trying it to upgrade an old (but relatively complex) build pipeline, and would like to pass a bunch of state known when creating the Parcel instance to a custom babel transform. Seems quite a waste (if not impossible?) to re-figure out these parameters in the transform. (But I've just started looking into Parcel, so let me know if it has a very good reason to only support paths, eg perhaps they serialize better than closures when passing to workers or so).

**Comment by CyriacBr** [comment]
> @padmaia Hmm, is dropping support for programmatically passing config objects intentional? I'm giving Parcel v2 a spin trying it to upgrade an old (but relatively complex) build pipeline, and would like to pass a bunch of state known when creating the Parcel instance to a custom babel transform. Seems quite a waste (if not impossible?) to re-figure out these parameters in the transform. (But I've just started looking into Parcel, so let me know if it has a very good reason to only support paths, eg perhaps they serialize better than closures when passing to workers or so).

Very surprised to see this removed as well! Programmatically defining a config should be an option like all other bundlers here that offer a programmatic API.

**Pull request merged**